File: src/main/scala/toolbox/Native.scala
package toolbox

import org.scalajs.dom.Blob
import scala.scalajs.js
import js.annotation.*

object Native {
  @js.native
  @JSGlobal("helloJS")
  def helloJS(): String = js.native

  @js.native
  @JSGlobal("downloadBlob")
  def downloadBlob(blob: Blob, name: String): Unit = js.native

  @js.native
  @JSGlobal("parseCsv")
  def parseCsv(text: String): js.Array[js.Array[String]] = js.native

  @js.native
  @JSGlobal("makeCsv")
  def makeCsv(rows: js.Array[js.Array[String]]): String = js.native

  @js.native
  @JSGlobal("copyToClipboard")
  def copyToClipboard(id: String): Unit = js.native
}

File: docs/native.js
function helloJS() {
  return "Hello Javascript!"
}

function downloadBlob(blob, name) {
  const blobUrl = URL.createObjectURL(blob);
  const link = document.createElement("a");
  link.href = blobUrl;
  link.download = name;
  document.body.appendChild(link);
  const event = new MouseEvent('click', { bubbles: true, cancelable: true, view: window});
  link.dispatchEvent(event);
  document.body.removeChild(link);
}

function parseCsv(content) {
  const csv = Papa.parse(content.trim());
  return csv.data;
}

function makeCsv(rows) {
  return Papa.unparse(rows);
}

function copyToClipboard(id) {
  var r = document.createRange();
  r.selectNode(document.getElementById(id));
  window.getSelection().removeAllRanges();
  window.getSelection().addRange(r);
  document.execCommand('copy');
  window.getSelection().removeAllRanges();
}
